### PR TITLE
Separate health checks by tags

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

See: https://github.com/ava-labs/avalanchego/discussions/1574

## How this works

- Marks failing health checks by tag
- Logs when a check goes from healthy -> unhealthy (WARN)
- Logs when a check goes from unhealthy -> healthy (INFO).

### Old

logs:
```
```

metrics:
```
# HELP avalanche_health_checks_failing number of currently failing health checks
# TYPE avalanche_health_checks_failing gauge
avalanche_health_checks_failing 0
# HELP avalanche_liveness_checks_failing number of currently failing health checks
# TYPE avalanche_liveness_checks_failing gauge
avalanche_liveness_checks_failing 0
# HELP avalanche_readiness_checks_failing number of currently failing health checks
# TYPE avalanche_readiness_checks_failing gauge
avalanche_readiness_checks_failing 0
```

### New

logs:
```
[06-02|17:24:54.604] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "database", "tags": ["application"]}
[06-02|17:24:54.604] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "router", "tags": ["application"]}
[06-02|17:24:54.604] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "diskspace", "tags": ["application"]}
[06-02|17:25:24.604] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "X", "tags": ["11111111111111111111111111111111LpoYY"]}
[06-02|17:25:24.610] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "network", "tags": ["application"]}
[06-02|17:25:24.611] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "C", "tags": ["11111111111111111111111111111111LpoYY"]}
[06-02|17:25:24.611] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "P", "tags": ["11111111111111111111111111111111LpoYY"]}
[06-02|17:26:24.606] INFO health/worker.go:235 check started passing {"namespace": "readiness", "name": "bootstrapped", "tags": ["application"]}
[06-02|17:26:24.606] INFO health/worker.go:235 check started passing {"namespace": "health", "name": "bootstrapped", "tags": ["application"]}
```
metrics:
```
# HELP avalanche_health_checks_failing number of currently failing health checks
# TYPE avalanche_health_checks_failing gauge
avalanche_health_checks_failing{tag="11111111111111111111111111111111LpoYY"} 0
avalanche_health_checks_failing{tag="all"} 0
avalanche_health_checks_failing{tag="application"} 0
# HELP avalanche_liveness_checks_failing number of currently failing health checks
# TYPE avalanche_liveness_checks_failing gauge
avalanche_liveness_checks_failing{tag="all"} 0
avalanche_liveness_checks_failing{tag="application"} 0
# HELP avalanche_readiness_checks_failing number of currently failing health checks
# TYPE avalanche_readiness_checks_failing gauge
avalanche_liveness_checks_failing{tag="all"} 0
avalanche_readiness_checks_failing{tag="application"} 0
```

### CODE REVIEW

**Comment by aaronbuchwald** [comment]
Tested locally and got from:

```
# TYPE avalanche_health_checks_failing gauge
avalanche_health_checks_failing{tag="11111111111111111111111111111111LpoYY"} 3
avalanche_health_checks_failing{tag="global"} 4
```

to

```
# HELP avalanche_health_checks_failing number of currently failing health checks
# TYPE avalanche_health_checks_failing gauge
avalanche_health_checks_failing{tag="11111111111111111111111111111111LpoYY"} 0
avalanche_health_checks_failing{tag="global"} 0
# HELP avalanche_liveness_checks_failing number of currently faili
```

nice 🤝

**Review by aaronbuchwald — DISMISSED** [review]

File: chains/manager.go [hunk 1]
```diff
@@ -963,7 +964,7 @@ func (m *manager) createAvalancheChain(
 	})
 
 	// Register health check for this chain
-	if err := m.Health.RegisterHealthCheck(chainAlias, h, ctx.SubnetID.String()); err != nil {
+	if err := m.Health.RegisterHealthCheck(chainAlias, h, health.GlobalTag, ctx.SubnetID.String()); err != nil {
```
Comment (ceyonur) [inline chains/manager.go:967]: same comment about showing all chains for filtered results

File: chains/manager.go [hunk 2]
```diff
@@ -364,6 +364,7 @@ func (m *manager) createChain(chainParams ChainParameters) {
 			health.CheckerFunc(func(context.Context) (interface{}, error) {
 				return nil, healthCheckErr
 			}),
+			health.GlobalTag,
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply chains/manager.go:367]: I think I forgot what `Global` was for... will address

File: api/health/metrics.go [hunk 3]
```diff
@@ -7,16 +7,20 @@ import "github.com/prometheus/client_golang/prometheus"
...
+		failingChecks: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "checks_failing",
+				Help:      "number of currently failing health checks",
+			},
+			[]string{"tag"},
+		),
 	}
+	metrics.failingChecks.WithLabelValues(GlobalTag).Set(0)
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply api/health/metrics.go]: renamed to `application`

File: api/health/worker.go [hunk 4]
```diff
@@ -12,41 +12,71 @@ import (
...
+	tags                        map[string]set.Set[string] // tag -> set of check names
 
 	startOnce sync.Once
 	closeOnce sync.Once
 	wg        sync.WaitGroup
 	closer    chan struct{}
 }
 
-func newWorker(namespace string, registerer prometheus.Registerer) (*worker, error) {
+type checker struct {
```
Comment (danlaine) [inline api/health/worker.go:50]: I'm not super happy about having a `checker` struct that has a `checker Checker` field so if you've got a better naming idea I'd be very amenable lol

Comment (danlaine) [reply api/health/worker.go:50]: Maybe this could be like, `taggedChecker` or something?

File: api/health/worker.go [hunk 5]
```diff
@@ -204,10 +265,41 @@ func (w *worker) runCheck(ctx context.Context, wg *sync.WaitGroup, name string,
...
+			if check.isApplicationCheck {
+				for tag := range w.tags {
+					w.metrics.failingChecks.WithLabelValues(tag).Inc()
+				}
+				w.numFailingApplicationChecks++
+			} else {
+				for _, tag := range check.tags {
+					w.metrics.failingChecks.WithLabelValues(tag).Inc()
+				}
+				w.metrics.failingChecks.WithLabelValues(AllTag).Inc()
```
Comment (danlaine) [inline api/health/worker.go:283]: Why don't we do this in the `if` statement? Don't we want to increment the total number of failing checks ?

Comment (danlaine) [reply api/health/worker.go:283]: Similar for decrementing this in the `else if prevResult.Error != nil` block below

**Review by danlaine — COMMENTED** [review]
a few questions

File: api/health/worker.go [hunk 6]
```diff
@@ -204,10 +265,41 @@ func (w *worker) runCheck(ctx context.Context, wg *sync.WaitGroup, name string,
 		}
 
 		if prevResult.Error == nil {
```
Comment (ceyonur) [inline api/health/worker.go:267]: generally this function has become a bit hard to follow with those ifs/elses

**Review by ceyonur — COMMENTED** [review]
changes LGTM (apart from the @danlaine's comment)

**Review by danlaine — APPROVED** [review]

File: api/health/worker.go (hunk 6 above)
Comment (StephenButtolph) [reply api/health/worker.go:267]: definitely open to suggestions... I couldn't find a cleaner way to do it

File: api/health/worker.go (hunk 4 above)
Comment (StephenButtolph) [reply api/health/worker.go]: sgtm

File: api/health/worker.go (hunk 5 above)
Comment (StephenButtolph) [reply api/health/worker.go:285]: added comments explaining why we don't do it in the `if`

File: api/health/worker.go (hunk 6 above)
Comment (ceyonur) [reply api/health/worker.go:268]: not sure if this is simpler, but IMO at least complexity reduced down to a helper function lol:
https://github.com/ava-labs/avalanchego/pull/1596/files
